lib/evaluate/type.h

```diff
@@ -35,168 +38,245 @@ namespace Fortran::evaluate {
+using ComplexTypes = CategoryTypes<TypeCategory::Complex>;
+using CharacterTypes = CategoryTypes<TypeCategory::Character>;
+using LogicalTypes = CategoryTypes<TypeCategory::Logical>;
+
+using FloatingTypes = common::CombineTuples<RealTypes, ComplexTypes>;
+using NumericTypes = common::CombineTuples<IntegerTypes, FloatingTypes>;
+using RelationalTypes = common::CombineTuples<NumericTypes, CharacterTypes>;
+using AllIntrinsicTypes = common::CombineTuples<RelationalTypes, LogicalTypes>;
+
+// When Scalar<T> is S, then TypeOf<S> is T.
+// TypeOf is implemented by scanning all supported types for a match
+// with Type<T>::Scalar.
+template<typename CONST> struct TypeOfHelper {
+  template<typename T> struct Predicate {
+    static constexpr bool value() {
+      return std::is_same_v<std::decay_t<CONST>,
+          std::decay_t<typename T::Scalar>>;
+    }
+  };
+  static constexpr int index{
+      common::SearchMembers<Predicate, AllIntrinsicTypes>};
+  using type = std::conditional_t<index >= 0,
+      std::tuple_element_t<index, AllIntrinsicTypes>, void>;
+};
+
+template<typename CONST> using TypeOf = typename TypeOfHelper<CONST>::type;
+
+// A variant union that can hold a scalar constant of any type chosen from
+// a set of types, which is passed in as a tuple of Type<> specializations.
+template<typename TYPES> struct SomeScalar {
+  using Types = TYPES;
+  CLASS_BOILERPLATE(SomeScalar)
+
+  template<typename A> SomeScalar(const A &x) : u{x} {}
   template<typename A>
-  GenericScalar(std::enable_if_t<!std::is_reference_v<A>, A> &&x)
+  SomeScalar(std::enable_if_t<!std::is_reference_v<A>, A> &&x)
     : u{std::move(x)} {}
 
-  std::optional<std::int64_t> ToInt64() const;
-  std::optional<std::string> ToString() const;
+  auto ToInt64() const {
+    return std::visit(
+        [](const auto &x) -> std::optional<std::int64_t> {
+          if constexpr (TypeOf<decltype(x)>::category ==
+              TypeCategory::Integer) {
+            return {x.ToInt64()};
+          }
+          return std::nullopt;
+        },
+        u);
+  }
 
-  std::variant<ScalarConstant<TypeCategory::Integer>,
-      ScalarConstant<TypeCategory::Real>, ScalarConstant<TypeCategory::Complex>,
-      ScalarConstant<TypeCategory::Character>,
-      ScalarConstant<TypeCategory::Logical>>
-      u;
+  auto ToString() const {
+    return std::visit(
+        [](const auto &x) -> std::optional<std::string> {
+          if constexpr (std::is_same_v<std::string,
+                            std::decay_t<decltype(x)>>) {
+            return {x};
+          }
+          return std::nullopt;
+        },
+        u);
+  }
+
+  auto IsTrue() const {
+    return std::visit(
+        [](const auto &x) -> std::optional<bool> {
+          if constexpr (TypeOf<decltype(x)>::category ==
+              TypeCategory::Logical) {
+            return {x.IsTrue()};
+          }
+          return std::nullopt;
+        },
+        u);
+  }
+
+  template<typename T> auto GetIf() const {
```

Comment:
Especially this one, where the return type isn't even in the function body.

Reply:
Deleting this function anyway after earlier review comment.